### PR TITLE
Add C++ syntax support for user-defined literals

### PR DESCRIPTION
**Pull request by adah1972**

### CODE REVIEW

**Comment by adah1972** [comment]
This PR makes Vim recognize user-defined literals like `"hello"sv`, `5.0i`, `300ms`, `0x123456789abcdef_cppi`, etc.

There are some repetitions, which seem usual in syntax files. However, if there is a better way of achieve such effects, let me know.

**Comment by mattn** [comment]
Looks good.

File: syntax/cpp.vim [hunk 1]
```diff
@@ -44,22 +44,33 @@ if !exists("cpp_no_cpp11")
...
   syn case match
 endif
 
 " C++ 20 extensions
 if !exists("cpp_no_cpp20")
+  syn match cppNumber		display "\<0b[01]\('\=[01]\+\)*\(y\|d\)\>"
+  syn match cppNumber		display "\<0\o\+\(y\|d\)\>"
+  syn match cppNumber		display "\<0\(y\|d\)\>"
+  syn match cppNumber		display "\<[1-9]\('\=\d\+\)*\(y\|d\)\>"
+  syn match cppNumber		display "\<0x\x\('\=\x\+\)*\(y\|d\)\>"
```
Comment (mattn) [inline syntax/cpp.vim:73]: Thank you. What specification this is numbers?

Comment (adah1972) [reply syntax/cpp.vim:73]: https://en.cppreference.com/w/cpp/chrono/operator%22%22y
https://en.cppreference.com/w/cpp/chrono/operator%22%22d

C++20 *years* and *days*.

**Comment by adah1972** [comment]
I have found another problem with case. I'll push a fix soon.

**Comment by adah1972** [comment]
Any more questions? Or more changes that I should make?

@mattn @k-takata

**Comment by mattn** [comment]
It seems okay to me. @k-takata what do you think this?

**Comment by mattn** [comment]
syntax for `0b0.1` is right?

image

**Comment by adah1972** [comment]
> syntax for `0b0.1` is right?

No, it is not. It is actually an interaction problem with c.vim (the fraction part is cFloat, not cppFloat).

I’ve pushed a fix. Also added the support for C++17 hexadecimal float literals.

**Comment by mattn** [comment]
Looks ok to me.

**Comment by k-takata** [comment]
LGTM

**Comment by adah1972** [comment]
Is there anything else I should do before merging?

**Comment by mattn** [comment]
I'll merge this and I'll send patch to vim-dev. Thanks..

**Comment by mattn** [comment]
https://groups.google.com/g/vim_dev/c/QiILCWOXnuY/m/UnWeCwmEBQAJ

**Comment by brammool** [comment]
I'll include the changes, thanks.

**Pull request merged**